students/pkoleyni/Lesson03/mailroom.py

```diff
@@ -121,17 +126,46 @@ def average_gifts(name):
             avarage = sum(donor_donation[1:]) / len(donor_donation[1:])
     return avarage
 
+
+def sort_key(key):
+    """
+
+    :param key: Key defines the item in the list to be used for sorting
+    :return:
+    """
+    return key[1]
+
 def report():
     """
 
     :return: will call other function to get the data and print the report
     """
+    donor_list_length = len(list_donors())
+    report_list = [[] for i in range(donor_list_length)]
     print()
+
+
+    for i in range(0,len(list_donors())):
+        donor_name = list_donors()[i]
+        report_list[i].append(donor_name)
+        report_list[i].append(total_donation(donor_name))
+        report_list[i].append(num_gifts(donor_name))
+        report_list[i].append(average_gifts(donor_name))
```

Comment:
I'm a bit confused on the above code, here is what i see:
* first loop on line 143 via call to list_donors()
* second loop on 144 to generate empty list of lists
* another loop on line 148 to call list_donors() (now inside of a for loop, so called each list item)
* then 3 more additional loops via total_donation(), num_gifts(), average_gifts()


above is all very expensive, you are looping over your list way too many times.

This should all be done in a single loop:
```
for donor in donor_list:  # loop over items in a list, don't use range()
    name = donor[0]
    donations = donor[1:]
    num_gifts = len(donations)
    total_donations = sum(donations)
    avg_gifts = total_donation / num_gifts
    report_list.append([name, total_donations, num_gifts, avg_gifts)
```

please review above code and let me know if you have any questions
    report_list.append([


```

Reply:
Thank you for the comment, I see that i was doing many unnecessary steps.
I have update the code 
https://github.com/pkoleyni/SP_Online_PY210/commit/87b4491cbc03dbe053214a3a6ec71faf31d9b01c

Reply:
sounds great!